crates/deltalake-core/src/operations/write.rs

```diff
@@ -940,4 +1118,168 @@ mod tests {
 
         assert_batches_eq!(&expected, &data);
     }
+
+    #[tokio::test]
+    async fn test_replace_where() {
+        let schema = get_arrow_schema(&None);
+
+        let batch = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(arrow::array::StringArray::from(vec!["A", "B", "C", "C"])),
+                Arc::new(arrow::array::Int32Array::from(vec![0, 20, 10, 100])),
+                Arc::new(arrow::array::StringArray::from(vec![
+                    "2021-02-02",
+                    "2021-02-03",
+                    "2021-02-02",
+                    "2021-02-04",
+                ])),
+            ],
+        )
+        .unwrap();
+
+        let table = DeltaOps::new_in_memory()
+            .write(vec![batch])
+            .with_save_mode(SaveMode::Append)
+            .await
+            .unwrap();
+        assert_eq!(table.version(), 0);
+
+        let batch_add = RecordBatch::try_new(
+            Arc::clone(&schema),
+            vec![
+                Arc::new(arrow::array::StringArray::from(vec!["C"])),
+                Arc::new(arrow::array::Int32Array::from(vec![50])),
+                Arc::new(arrow::array::StringArray::from(vec!["2023-01-01"])),
+            ],
+        )
+        .unwrap();
+
+        let table = DeltaOps(table)
+            .write(vec![batch_add])
+            .with_save_mode(SaveMode::Overwrite)
+            .with_replace_where(col("id").eq(lit("C")))
+            .await
+            .unwrap();
+        assert_eq!(table.version(), 1);
+
+        let expected = [
+            "+----+-------+------------+",
+            "| id | value | modified   |",
+            "+----+-------+------------+",
+            "| A  | 0     | 2021-02-02 |",
+            "| B  | 20    | 2021-02-03 |",
+            "| C  | 50    | 2023-01-01 |",
+            "+----+-------+------------+",
+        ];
+        let actual = get_data(&table).await;
+        assert_batches_sorted_eq!(&expected, &actual);
+    }
+
+    #[tokio::test]
+    async fn test_replace_where_fail_not_matching_predicate() {
+        // Using a temp file-based table instead of `memory://` table because we need to reload the table
+        // after a failed write attempt, which seem to be impossible for memory tables
```

Comment:
It looks like something that didn't work few weeks ago works now so I'll refactor

TLDR;

After L1222-L1227 the table is in error state so I need to reload it to check it's current state. I do that on L1229 but if I use `memory://` I get a `NotATable("no log files")` error. 

I tried cloning the table's `state` and `logstore` before L1222 but that was causing a different error when I was trying to read the table in L1230. I wanted to get the exact error just now, tried it again and ... it works. Something may have changed since my last try because I am using the same code that didn't work few weeks ago.